File: frontend/src/types/recap.ts
export interface LeaderboardEntry {
  username: string
  count: number
}

export interface RecapAlbumCard {
  album_id: number
  spotify_album_id: string | null
  title: string
  artist: string | null
  cover_url: string | null
  avg_rating: number
  review_count: number
  weighted_score: number
}

export interface MemberGuessAccuracy {
  username: string
  total: number
  correct: number
  pct: number
}

export interface GuessAccuracy {
  total_guesses: number
  correct_guesses: number
  pct: number
  per_member: MemberGuessAccuracy[]
}

export interface RecapData {
  albums_added: LeaderboardEntry[]
  albums_reviewed: LeaderboardEntry[]
  favorite_album: RecapAlbumCard | null
  least_favorite_album: RecapAlbumCard | null
  guess_accuracy: GuessAccuracy
}

export interface RecapResponse {
  id: number
  group_id: number
  week_start: string
  week_end: string
  generated_at: string
  data: RecapData
  seen: boolean
}

export interface RecapSummary {
  id: number
  group_id: number
  group_name: string
  week_start: string
  week_end: string
}

File: backend/app/schemas/recap.py
"""Schemas for the weekly group recap.

``RecapData`` is the frozen payload persisted in ``GroupRecap.data``; the other
models wrap it for API responses and the login pop-up.
"""

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class LeaderboardEntry(BaseModel):
    username: str
    count: int


class RecapAlbumCard(BaseModel):
    album_id: int
    spotify_album_id: str | None = None
    title: str
    artist: str | None = None
    cover_url: str | None = None
    avg_rating: float
    review_count: int
    weighted_score: float


class MemberGuessAccuracy(BaseModel):
    username: str
    total: int
    correct: int
    pct: float


class GuessAccuracy(BaseModel):
    total_guesses: int
    correct_guesses: int
    pct: float
    per_member: list[MemberGuessAccuracy]


class RecapData(BaseModel):
    """The immutable computed payload stored on a recap."""

    albums_added: list[LeaderboardEntry]
    albums_reviewed: list[LeaderboardEntry]
    favorite_album: RecapAlbumCard | None
    least_favorite_album: RecapAlbumCard | None
    guess_accuracy: GuessAccuracy


class RecapResponse(BaseModel):
    id: int
    group_id: int
    week_start: date
    week_end: date
    generated_at: datetime
    data: RecapData
    seen: bool

    model_config = ConfigDict(from_attributes=True)


class RecapSummary(BaseModel):
    """Lightweight recap descriptor for the pending-recaps pop-up and lists."""

    id: int
    group_id: int
    group_name: str
    week_start: date
    week_end: date
